game-core/src/main/java/games/strategy/engine/framework/map/download/ContentDownloader.java

```diff
@@ -0,0 +1,79 @@
+package games.strategy.engine.framework.map.download;
+
+import games.strategy.engine.framework.system.HttpProxy;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Consumer;
+import lombok.Getter;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+
+/**
+ * Downloads content from HTTP resources. Meant to be used with a try-with-resources block, eg:
+ *
+ * <pre>
+ *   try(CloseableDownloader downloader = new ContentDownloader(uri)) {
+ *     InputStream stream = downloader.getStream();
+ *     // read and process stream, stream is auto-closed for you.
+ *   }
+ * </pre>
+ *
+ * To process and read the input stream in one-go, use {@see ContentReader}. Use this class if you
+ * want direct access to an InputStream. This is useful if you want to be able to change the source
+ * of the input stream.
+ *
+ * <p>Warning: The input stream provided by this class can only be consumed once (property of input
+ * streams, the input stream is not reset in any way after reading it).
+ */
+public final class ContentDownloader implements CloseableDownloader {
+  private final CloseableHttpClient httpClient;
+
+  @Getter(onMethod_ = @Override)
+  private final InputStream stream;
+
+  private final CloseableHttpResponse response;
+
+  public ContentDownloader(final URI uri) throws IOException {
+    this(HttpClients.custom().disableCookieManagement().build(), uri, HttpProxy::addProxy);
```

Comment:
I haven't used Feign for generic HTTP, always for JSON response, but I suppose nothing would stop it.

The `HttpClient` linked would be an upgrade to the code that was modified in this class, having 3 different objects to close and the `getEntity()` to return an `InputStream` all seem a bit unwieldy.

There is a question whether to migrate or not. I would be concerned about 'lava-flow' and code churn. I'm not sure if the ROI is really that great. This HTTP client would benefit, and we have the 'forum-poster' that is now inconsistent. The 'forum-poster' client I've thought would be better off as a feign client. I think for consistency, because of the API it operates with is fundamentally JSON based, I'd vote for now we use Feign for the JSON API's consistently. Otherwise, someone would have to evaluate the pros/cons of migrating away from Feign and hopefully be prepared to make an across-the-board update. I think ROI comes into play though, a Reactive interface is nice, Feign interfaces are nice, IMO we are perhaps in a good-enough place and maybe better to focus on other things and avoid the churn. With all that said, refactoring `ContentDownloader` to use the reactive API or feign would be good IMO.

Reply:
Just tried migrating away from the websocket lib were using, and whilst definitely feasible it's currently not worth my time: The standard lib has a completely different API structure, which would require us to change a lot of the class structure